_posts/2016-08-22-the-java-synchronisers.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: post
+asset-type: post
+name: the-java-synchronisers
+title: 'The Java Syncrhonisers' 
+date: 2016-08-22 10:00:00 +00:00
+author: Carlos Morera de la Chica
+image:
+    src: /assets/img/custom/blog/synchronisation.jpg
+tags:
+- java
+- threading
+- synchronization
+
+canonical:
+    name: my personal blog
+    href: https://carlosmchica.github.io/the-java-synchronisers/
+
+---
+
+Threads communication happens primarily by sharing access to fields and objects. Although extremely efficient, this form of communication can lead to errors such as [thread interference](https://docs.oracle.com/javase/tutorial/essential/concurrency/interfere.html) and [memory consistency](https://docs.oracle.com/javase/tutorial/essential/concurrency/memconsist.html). Synchronization is a tool that helps to prevent these errors.
+
+However, synchronization does not come for free and can introduce waitings to access a lock or object that is currently being held by another thread. The waiting thread cannot use that object until the other thread releases the object. This condition is known as thread contention.
+
+In this post we will explore the different options that Java provides to deal with threads synchronization.
+
+## Synchronization essentials
+
+Java provides a series of mechanisms to handle basic thread synchronization and coordination. It supports fine-grained object access synchronization through [synchronized methods](#sync-methods) and [synchronized statements](#sync-statements). Basic thread coordination can be accomplished through [guarded blocks](#guarded-blocks). All mentioned mechanisms are built around acquiring and releasing objects' intrinsic lock.
+
+### Intrisic Lock
+
+Every Java object has associated an intrinsic lock. A thread that needs exclusive access to an object's fields has to acquire the object's lock before accessing them, releasing the intrinsic lock once it is done. Other threads trying to access the object will block until the thread holding the lock releases it.
+
+### <a name="sync-methods"></a> Synchronized methods
+
+When a thread invokes a [synchronized method](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html), it acquires the **intrinsic lock** for that method's object and releases it when the methods returns. The lock is released even if the method returns due to an uncaught exception. If done in a static method, the thread acquires the lock for the `class` object associated with the class.
```

Comment:
`class` object, here show what you mean by it, its usually the `Xxxxx.class` object - see if there is a reference to it and link it here.